internal/speakerlist/speakerlist.go

```diff
@@ -97,193 +91,164 @@ func New(logger log.Logger, nodeName, bindAddr, bindPort, secret, namespace, lab
-	go sl.updateSpeakerIPs()
-
+	sl.mlJoinCh = make(chan struct{}, 1)
+	sl.stopCh = make(chan struct{})
 	go sl.memberlistWatchEvents()
 	go sl.joinMembers()
+
+	return &sl, nil
 }
 
-// updateSpeakerIPs runs forever updating the sl.mlSpeakerIPs slice with the
-// IPs of all the speaker pods. As the function queries the API server, it waits at
-// least 5 minutes between queries to avoid self-induced API server DoS (should
-// be treated with care).
-func (sl *SpeakerList) updateSpeakerIPs() {
+func (sl *SpeakerList) joinMembers() {
 	for {
-		// This either stops the loop or makes sure to wait at least 5
-		// minutes before making another call to sl.mlSpeakers(), which
-		// queries the API server.
 		select {
+		case <-sl.mlJoinCh:
+			sl.Lock()
+			joinList := sl.mlJoinList
+			sl.Unlock()
+			if len(joinList) == 0 {
+				// joinList is empty - wait for a new rejoin() call.
+				break
+			}
+			// Join is sequential and TCPTimeout = 10s, so it can take time.
+			n, err := sl.ml.Join(joinList)
+			if err == nil {
+				level.Info(sl.l).Log("op", "ml-join", "msg", "memberlist join", "no. joined", n, "join list", joinList)
+			} else {
+				level.Error(sl.l).Log("op", "ml-join", "msg", "memberlist join", "no. joined", n, "error", err, "join list", joinList)
+			}
+			select {
+			case <-time.After(15 * time.Second):
+				// Wait 15s between join attemps, by calling rejoin() we go to 'case <-sl.mlJoinCh',
+				// so if mlJoinList is now empty we will break.
+				sl.rejoin()
+			case <-sl.stopCh:
+				return
+			}
 		case <-sl.stopCh:
 			return
-		case <-time.After(5 * time.Minute):
-			// This blocks until the API server responds.
-			iplist, err := sl.mlSpeakers()
-			if err != nil {
-				level.Error(sl.l).Log("op", "memberDiscovery", "error", err, "msg", "failed to get pod IPs")
-				continue
-			}
-
-			sl.mlMux.Lock()
-			sl.mlSpeakerIPs = iplist
-			sl.mlMux.Unlock()
 		}
 	}
 }
 
-func (sl *SpeakerList) mlSpeakers() ([]string, error) {
-	// This call blocks until we get a response from the API server.
-	// In the client-go version we are using, there is no way to use a
-	// context. Newer versions of client-go support using a context
-	// in the call sl.client.PodIPs() is using.
-	//
-	// TODO: When updating client-go, this code can be simplified by using a
-	// context in the following way:
-	// If we use a context, we can specify a timeout and then there is no need to
-	// run mlUpdateSpeaker() in the background. We could then call
-	// this function from mlJoin() with a context (timeout) and use
-	// sl.mlSpeakerIPs on failure. On success, we could update
-	// sl.mlSpeakerIPs with the response and continue to do the join.
-	iplist, err := sl.client.PodIPs(sl.namespace, sl.labels)
-	if err != nil {
-		return nil, err
+// updateJoinList updates mlJoinList with the IPs of all the ready speakers
+// (ie present in k8sSpeakers) that are not alive members of the memberlist cluster
+// (ie not present in mlMembers).
```

Comment:
```suggestion
// (i.e. present in k8sSpeakers) that are not alive members of the memberlist cluster
// (i.e. not present in mlMembers).
```

Reply:
godot is not smart enough :(
```
internal/speakerlist/speakerlist.go:139:10: Sentence should start with a capital letter (godot)
// (i.e. present in k8sSpeakers) that are not alive members of the memberlist cluster
         ^
internal/speakerlist/speakerlist.go:140:10: Sentence should start with a capital letter (godot)
// (i.e. not present in mlMembers).
         ^

Exited with code exit status 1
```

Reply:
This bug was fixed in https://github.com/tetafro/godot/commit/09b0ef39f22fe608d098de46c5efa7b88c5cdccb and updated in golangci in https://github.com/golangci/golangci-lint/pull/2123. Until a new golangci version is released, feel free to disable godot using `//nolint:godot` in the relevant places.

Reply:
done